### PR TITLE
InputDialog and MultiInputDialog: improve enter and edit callback handling, and implement addWidget

### PR DESCRIPTION
**Pull request by ziz57**
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/12289)

### CODE REVIEW

File: frontend/ui/widget/inputdialog.lua [hunk 1]
```diff
@@ -336,6 +336,16 @@ function InputDialog:init()
             self._top_line_num, self._charpos = top_line_num, charpos
         end
     end
+    self.enter_callback = self.enter_callback or function()
+            for _,btn_row in ipairs(self.buttons) do
+                for _,btn in ipairs(btn_row) do
+                    if btn.is_enter_default then
```
Comment (poire-z) [inline frontend/ui/widget/inputdialog.lua:342]: Too much indentation on the first "for".
Also, it was already bad in the original code you moved here, but please add the missing space after the comma: `_, btn`

File: frontend/ui/widget/inputdialog.lua [hunk 2]
```diff
@@ -352,18 +362,9 @@ function InputDialog:init()
...
-                        btn.callback()
-                        return
-                    end
-                end
-            end
-        end,
+        enter_callback = not self.allow_newline and self.enter_callback,
         strike_callback = self.strike_callback,
-        edit_callback = self._buttons_edit_callback, -- nil if no Save/Close buttons
+        edit_callback = self._buttons_edit_callback or self.edited_callback,
```
Comment (poire-z) [inline frontend/ui/widget/inputdialog.lua:367]: Not accurate anymore to keep that comment ?

File: frontend/ui/widget/multiinputdialog.lua [hunk 3]
```diff
@@ -249,4 +265,24 @@ function MultiInputDialog:onKeyboardHeightChanged()
...
+                h = widget:getSize().h,
+            },
+            widget,
+        }
+        if not self._added_widgets then
+            self._added_widgets = {}
+        end
+        table.insert(self._added_widgets, widget)
+    end
+    -- insert widget before the bottom buttons
```
Comment (poire-z) [inline frontend/ui/widget/multiinputdialog.lua:283]: The origina you copied from inputdialog.lua has:
`-- insert widget before the bottom buttons and their previous vspan`
Not exactly the case in here ? If it is, please just keep them exactly similar (so we don't have to wonder if touching this later).

**Review by poire-z — COMMENTED** [review]
Fine with me, just some nitpicks.

File: frontend/ui/widget/multiinputdialog.lua [hunk 4]
```diff
@@ -249,4 +265,24 @@ function MultiInputDialog:onKeyboardHeightChanged()
     UIManager:setDirty("all", "flashui")
 end
 
+function InputDialog:addWidget(widget, re_init)
```
Comment (hius07) [inline frontend/ui/widget/multiinputdialog.lua:268]: Base class?

Comment (poire-z) [reply frontend/ui/widget/multiinputdialog.lua:268]: Nice catch, you have good eyes :)

File: frontend/ui/widget/inputdialog.lua (hunk 1 above)
Comment (ziz57) [reply frontend/ui/widget/inputdialog.lua]: Fixed.

File: frontend/ui/widget/inputdialog.lua (hunk 2 above)
Comment (ziz57) [reply frontend/ui/widget/inputdialog.lua]: Fixed.

File: frontend/ui/widget/multiinputdialog.lua (hunk 3 above)
Comment (ziz57) [reply frontend/ui/widget/multiinputdialog.lua]: Fixed (it is indeed directly analogous)

**Review by poire-z — APPROVED** [review]

**Review by Frenzie — APPROVED** [review]

**Review by hius07 — APPROVED** [review]

**Pull request merged**